Vulkan AS rebuild-on-replay: On queue submission completion events

cmannett85-arm opened the pull request:
If an AS build command is called on an already built AS (i.e. an update build) then we need to safely destroy the previously copied input buffers and their memory.  However at the point this happens, we do not own the command buffer so we don't know if the original data has actually been copied yet.

Using the mechanism in this patch we associate function objects with the command buffer that will be executed when the queue the command buffer is submitted to, has finished executing on the GPU.

This works by carying the function objects in command buffer and once it is submitted, we poll on a VkEvent we inserted.  Once signalled, we call the function objects in the order they were inserted.

baldurk reviewed (changes requested):
I would rather have a different solution for this which I think should be simpler/more direct and more tolerant of common application behaviours.

* We can't expect each submission to complete immediately (or within 5ms) so instead we should check after each submission to see which things we're waiting on have been completed, and add new things into that list. Any which have completed should be processed, any that haven't can continue waiting.
* Since we're not waiting immediately, we can't use the user's fence as we can't reference it after the submit call returns. In theory the ideal solution would be to add our own semaphore to the submission's signalled semaphore, but I don't see any way in base vulkan to check a semaphore's status from the CPU and although it might be safe to rely on timeline semaphores I'd rather not depend on another extension. Instead I think it's best to set a VkEvent in the command buffer and poll on that.
* Instead of divorcing the registration of 'callback after X command buffer completes' with its processing via a map lookup, store the callbacks to be called directly in the command buffer record and during queue submit we can move them out into the array of events we're waiting for. We can store a `{event, callback}` pair.
* Rather than adding a new class for this which adds another indirection and I find makes things harder to navigate/organise, keep this within `WrappedVulkan` or at most in `VulkanAccelerationStructureManager`. Having it in a separate file is fine if you want but I think if the above changes are made there's not much to go there.

If you look in the D3D12 implementation at `WrappedID3D12GraphicsCommandList::AddSubmissionASBuildCallback` through to `WrappedID3D12CommandQueue::ExecuteCommandListsInternal` you can see where this kind of callback is recorded in the command buffer, then taken by the queue and passed to the RT-specific manager for polling.

As a minor note - there's no need to store a VkDevice. We only handle a single device and that's unlikely to ever change in future, if it does there will need to be a pass done throughout the codebase to handle it.

cmannett85-arm commented:
Ah balls, this is wrong.  I'm calling `vkCmdSetEvent` on the command buffer after `vkEndCommandBuffer` has been called which bizarrely works on Mali but (quite fairly) causes the VVL to have a meltdown.

cmannett85-arm committed:
Vulkan AS rebuild-on-replay: On queue submission completion events

If an AS build command is called on an already built AS (i.e. an update build) then we need to safely destroy the previously copied input buffers and their memory.  However at the point this happens, we do not own the command buffer so we don't know if the original data has actually been copied yet.

Using the mechanism in this patch we associate function objects with the command buffer that will be executed when the queue the command buffer is submitted to, has finished executing on the GPU.

This works by carying the function objects in command buffer and once it is submitted, we poll on a VkEvent we inserted.  Once signalled, we call the function objects in the order they were inserted.

baldurk commented on renderdoc/driver/vulkan/vk_core.cpp:
```diff
@@ -703,6 +703,75 @@ void WrappedVulkan::SetDebugMessageSink(WrappedVulkan::ScopedDebugMessageSink *s
+
+void WrappedVulkan::AddPendingCommandBufferCallbacks(VkCommandBuffer commandBuffer)
+{
+  SCOPED_LOCK(m_PendingCmdBufferCallbacksLock);
```
This lock should be grabbed right before `m_PendingCmdBufferCallbacks.push_back` to minimise contention so that it's not held while calling into the driver.

baldurk commented on renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp:
```diff
@@ -1378,11 +1378,14 @@ VkResult WrappedVulkan::vkBeginCommandBuffer(VkCommandBuffer commandBuffer,
     // any existing baked commands before creating a new set.
     if(record->bakedCommands)
+    {
+      MarkPendingCommandBufferAsDeleted(commandBuffer);
```
This handling also needs to be added to `vkResetCommandBuffer`.

baldurk reviewed (changes requested):
I feel that this could be simplified a little further overall - the lifetime of the event and needing to track it vs the command buffer in two places makes this a little awkward.

In practice we could just delete the event as soon as we see it's set, but the spec disallows that until we can prove that the command buffer has finished executing and hooking onto the reset/free is a good plan for that. I even wondered if we could pool the events such that we don't have to worry about their lifetime but the spec also bans submitting a use of an event while another use of it is in flight - even though this would again be ordered in practice by the CPU check :(.

I think putting the command buffer more explicitly in control of the event's lifetime would work nicely - since you have one event per command buffer it seems to me to make more sense to tie them together. Then you only have to handle being able to check the event potentially after the command buffer has been reset.

I'd suggest adding a simple refcount for the event, have both the command buffer and the overall global array of callbacks holding a ref to it. Once both have release their ref it deletes itself. You could leave the native VkEvent to be entirely tied to the lifetime of the command buffer and simply have a helper refcounted struct which knows that a deleted event means implicitly that it's been set.

cmannett85-arm commented:
Apologies for the delay, I've just got back from being abroad.

> I'd suggest adding a simple refcount for the event, have both the command buffer and the overall global array of callbacks holding a ref to it.

Any objections to just using an equivalent to `std::shared_ptr<VkEvent>(event, [d](VkEvent *event){ ObjDisp(d)->DestroyEvent(Unwrap(d), *event, NULL); delete event; }` for this?  I can't see `std::shared_ptr` being used elsewhere in the codebase but it saves me reinventing the refcounting wheel.

baldurk commented:
I'm glad you were able to get away, that's nothing to apologise for :).

> Any objections to just using an equivalent to `std::shared_ptr`

No please don't use a `std::` class for this, you can do the same as `ResourceRecord` for its refcounting:

```
  int32_t RefCount = 1;
  void AddRef() { Atomic::Inc32(&RefCount); }
  void Release()
  {
    int32_t ref = Atomic::Dec32(&RefCount);
    RDCASSERT(ref >= 0);
    if(ref <= 0)
      delete this;
  }
```

cmannett85-arm commented:
I'm not convinced this is much of an improvement...  But it's a starting point for iteration!

baldurk commented on renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp:
```diff
@@ -1670,6 +1670,9 @@ VkResult WrappedVulkan::vkEndCommandBuffer(VkCommandBuffer commandBuffer)
   RDCASSERT(record);
 
+  if(IsCaptureMode(m_State))
+    AddPendingCommandBufferCallbacks(commandBuffer);
```
We shouldn't add these to the global list until the command buffer is submitted to the queue and we know it will eventually complete. Otherwise if the command buffer isn't submitted the event will never be set, and the reference will stick around forever.

Here you need to create & set the event but that should be done internally.

baldurk commented on renderdoc/driver/vulkan/vk_resources.h:
```diff
@@ -1071,6 +1071,47 @@ struct MemRefs
+    VkEvent event;
+  };
+
+  static Threading::CriticalSection m_Lock;
```
Having a global lock here across all objects is not the best way to implement refcounting as it will introduce contention between unrelated objects and is a heavy-weight solution to synchronisation. If you look at how `ResourceRecord` is implemented with the code snippet I posted before it uses the typical approach of atomic operations to count the refcount and delete when it reaches 0.

baldurk commented on renderdoc/driver/vulkan/vk_resources.h:
```diff
@@ -1071,6 +1071,47 @@ struct MemRefs
+
+  static Threading::CriticalSection m_Lock;
+
+  DPtr *d = NULL;
```
Rather than having a value-copyable wrapper around a pointer like this you can remove the internal hidden pointer and make the `VkPendingSubmissionCompleteCallbacks` itself dynamically allocated. It can track its own refcount and delete itself when it reaches 0 as with `ResourceRecord`.

baldurk commented on renderdoc/driver/vulkan/vk_core.cpp:
```diff
@@ -705,6 +705,59 @@ void WrappedVulkan::SetDebugMessageSink(WrappedVulkan::ScopedDebugMessageSink *s
+      for(std::function<void()> &f : pending.callbacks)
+        f();
+
+      m_PendingCmdBufferCallbacks.erase(i);
```
If you're doing the loop like this (which is reasonable) you need to add a `continue` here to re-check `i` which will now be the next item in the list. As it is this will skip the element after each one erased.

cmannett85-arm replied:
Yep, I used to have one and then accidently deleted it after some rearranging...

baldurk commented on renderdoc/driver/vulkan/vk_resources.h:
```diff
@@ -1071,6 +1071,47 @@ struct MemRefs
+  {
+    SCOPED_LOCK(m_Lock);
+    return d == NULL ? VK_NULL_HANDLE : d->event;
+  }
```
I think rather than having an implicit conversion operator the event should just be a public member for use like `callbacks`.

baldurk commented on renderdoc/driver/vulkan/vk_core.cpp:
```diff
@@ -705,6 +705,59 @@ void WrappedVulkan::SetDebugMessageSink(WrappedVulkan::ScopedDebugMessageSink *s
+
+  rdcarray<std::function<void()>> callbacks;
+  {
+    SCOPED_LOCK(m_PendingCmdBufferCallbacksLock);
```
I don't think you need to lock `m_PendingCmdBufferCallbacksLock` here? It should only be necessary below to guard `m_PendingCmdBufferCallbacks`.

cmannett85-arm replied:
Not it's not, I'll remove it.  Though will accesses to `cmdRecord->...->pendingSubmissionCompleteCallbacks` need synchronising?  I suspose it does as additions to it are triggered by `vkCmd*` calls which can be called from mutliple threads.

baldurk replied:
That should be safe based on the vulkan rules - only one thread can be recording to a command buffer at any given time.

baldurk commented on renderdoc/driver/vulkan/wrappers/vk_queue_funcs.cpp:
```diff
@@ -1043,6 +1043,9 @@ void WrappedVulkan::CaptureQueueSubmit(VkQueue queue,
       record->bakedCommands->AddRef();
     }
+
+    if(backframe || capframe)
```
nit: This check is redundant as we're always in either one of these states during capture.

baldurk commented on renderdoc/driver/vulkan/vk_core.cpp:
```diff
@@ -705,6 +705,78 @@ void WrappedVulkan::SetDebugMessageSink(WrappedVulkan::ScopedDebugMessageSink *s
+    return;
+
+  rdcarray<std::function<void()>> callbacks;
+  callbacks.swap(pending->callbacks);
```
I'm not sure I understand the reason for swapping these callbacks into a local array only to move them back to the pending array at the end of this function?

cmannett85-arm replied:
A hangover from a previous version, removed now.

baldurk commented on renderdoc/driver/vulkan/vk_core.cpp:
```diff
@@ -705,6 +705,78 @@ void WrappedVulkan::SetDebugMessageSink(WrappedVulkan::ScopedDebugMessageSink *s
+
+void WrappedVulkan::CheckPendingCommandBufferCallbacks()
+{
+  SCOPED_LOCK(m_PendingCmdBufferCallbacksLock);
```
This is fine right now and doesn't need to change, but it's worth having a mental pin in this that this function will hold the lock for quite a while depending on how long the `vkGetEventStatus` call takes, as well as the callbacks themselves.

If there's contention in the future this could be improved, but for now the only thing this would do is block one queue submission on one thread waiting to add to the array while another thread is submitting to a different queue and checking these.

cmannett85-arm replied:
A trivial change could be to:

1. Acquire the lock
2. Move `m_PendingCmdBufferCallbacks` into a local
3. Release the lock
4. Do the checks/execution
5. Acquire the lock
6. _Prepend_ any remaining entries to `m_PendingCmdBufferCallbacks` (which may have accumulated new entries from other threads)
7. Release the lock

If not I can leave this as a comment as inspiration for some future person.

baldurk replied:
Yep that would certainly work, I don't think you need to specifically prepend entries in step 6 since we don't rely on an order of them retiring?

It's fine to leave this for now as I don't think it's a problem, but it's worth being aware of in case this does cause issues in the future or bearing in mind for future code in this area.

baldurk reviewed (changes requested):
I think this looks about ready - I've left a couple of minor notes only.

cmannett85-arm committed:
Switch to a ref-counted solution
